### ISSUE TITLE
Packer does weird things with escape characters

### ISSUE THREAD

**Issue by ziggythehamster** [rprichard/winpty#104]
If I run [Packer](http://www.packer.io) under winpty, I get the following output:

![image](https://cloud.githubusercontent.com/assets/66674/23934534/43e71fb6-0903-11e7-9fdd-322632ee5246.png)

Here's what it looks like under a native cmd.exe:

![image](https://cloud.githubusercontent.com/assets/66674/23934550/596db8c2-0903-11e7-830e-f7bc5ab44931.png)

This is winpty 0.4.2 under MSYS2 under ConEmu (with the ConEmu connector). Other native applications that use color work fine under this configuration. In particular, Terraform (also by Hashicorp) works fine (`winpty terraform foo`). Oddly, if I run this without using winpty, color works fine, *but* Ctrl-C isn't captured correctly.

**Comment by rprichard** [comment]
I'm guessing Packer thinks it can write escape sequences for some reason, even though it's writing to the winpty hidden console (and the hidden console probably doesn't have the Win10 escape handling enabled).  Maybe Packer notices ConEmu somehow and thinks ConEmu is patching escaped color support into WriteConsole using its hook DLL.

What OS are you using?  If you're using Win10, do you have the "Use legacy console" checkbox checked?  (http://www.isunshare.com/windows-10/enable-use-legacy-console-for-command-prompt-in-windows-10.html)

**Comment by rprichard** [comment]
What version of packer are you using?  I'm testing with 0.12.3, and I'm unable to get it to print colored output on Windows, regardless of what environment (e.g. ConEmu, mintty, ordinary console, etc.) I use.

**Comment by rprichard** [comment]
My current impression is that this is a Packer bug.  There doesn't appear to be a Packer Cygwin package, so I'm assuming we're talking about native Packer executables.

If I understand correctly, Packer colorizes output by writing escape sequences to stdout, and it decides that these escapes are supported when this function returns true:

```
func (u *ColoredUi) supportsColors() bool {
	// Never use colors if we have this environmental variable
	if os.Getenv("PACKER_NO_COLOR") != "" {
		return false
	}

	// For now, on non-Windows machine, just assume it does
	if runtime.GOOS != "windows" {
		return true
	}

	// On Windows, if we appear to be in Cygwin, then it does
	cygwin := os.Getenv("CYGWIN") != "" ||
		os.Getenv("OSTYPE") == "cygwin" ||
		os.Getenv("TERM") == "cygwin"

	return cygwin
}
```

(https://github.com/mitchellh/packer/blob/b8cf6a9e1b4a5ac8565bf5bd3c0f96acf4b8f93c/packer/ui.go#L128)

I'm guessing there isn't a Go-language layer interpreting escape sequences.

I don't think the Windows part of the function makes any sense.  Here's what I see when I test from a Cygwin mintty Bash shell:

```
$ set | grep '\(CYGWIN\|OSTYPE\|TERM\)'
OSTYPE=cygwin
TERM=xterm

$ sh -c set | grep '\(CYGWIN\|OSTYPE\|TERM\)'
OSTYPE=cygwin
TERM=xterm

$ env | grep '\(CYGWIN\|OSTYPE\|TERM\)'
TERM=xterm

$ cmd /c set | grep '\(CYGWIN\|OSTYPE\|TERM\)'
TERM=xterm
```

The `OSTYPE` variable seems to be a shell variable, not an environment variable.  `CYGWIN` isn't normally set at all, and `TERM` is set to `xterm`, because mintty is a proper xterm.

`TERM=cygwin` indicates the limited terminal-to-console emulation layer that Cygwin uses when it's running in a Windows console.  However, this emulation layer only applies to *Cygwin* programs, and Packer isn't one.

I was able to enable the Packer color escapes by starting Cygwin in a console (e.g. Start->Run, `C:\cygwin64\bin\bash.exe --login`), then running Packer.  Because my console has no ANSI escape support (Win7), though, I see this:
```
rprichard@VBWIN7 ~/work/packer-tmp
$ packer build example.json
←[1;32mamazon-ebs output will be in this color.←[0m
←[1;36mamazon-ebs-2 output will be in this color.←[0m
```

I'm not sure why @ziggythehamster is seeing the broken escapes, because I'd expect `OSTYPE` and `CYGWIN` to be unset, and the ConEmu MSYS2 connector seems to set `TERM` to `xterm-256color`, not `cygwin`.  There are various odd reasons to set the `CYGWIN` flag, so maybe he set it.

A better check for Packer might be:

```
	// On Windows, if we appear to be in Cygwin, then it does
	cygwin := os.Getenv("TERM") != "" &&
		os.Getenv("TERM") != "dummy" &&     (or TERM=dumb, maybe?)
		os.Getenv("TERM") != "cygwin"
```

I don't recall exactly what `TERM=dummy` means, but I thought it meant something somewhere.  It could be left out, maybe.  Alternatively, Packer could look for common terminals (e.g. `xterm` and anything starting with `xterm-`, `rxvt`, etc.)

Packer could also look for the `ConEmuANSI=ON` environment variable.  ConEmu has a hook DLL that adds escape sequence support to `WriteConsole`/`WriteFile`, and it sets `ConEmuANSI` to signal that the hook DLL is present.

Packer could also see if the new Win10 console buffer mode flag, `ENABLE_VIRTUAL_TERMINAL_PROCESSING`, is enabled on the stdout console handle.  If it isn't enabled, it could try enabling it (but would it disable the flag on exit?  Yuck).  IIRC, some other Go program also supports this flag somehow.  Unfortunately, I think `cmd.exe` leaves the flag disabled by default.

To summarize, if I understand correctly, the existing check is broken because:
 * `OSTYPE` won't be set
 * `CYGWIN` is usually unset, but if it is, it'll have nothing to do with color support
 * `TERM=cygwin` doesn't indicate that Packer can output color -- it means exactly the opposite!

**Comment by ziggythehamster** [comment]
I think you nailed it - I have `CYGWIN=winsymlinks:native`. So, the reason that I get colors at all is because ConEmu is processing them... but only in a native cmd.exe under its control. And if I didn't have `CYGWIN` set, I wouldn't get colors either.

Terraform seems to work 100% correctly under winpty, which is also written in Go, so it sounds like Hashicorp should abstract their console code from Terraform and use it in both Packer and Terraform.

**Comment by rprichard** [comment]
> I think you nailed it - I have CYGWIN=winsymlinks:native. So, the reason that I get colors at all is because ConEmu is processing them... but only in a native cmd.exe under its control. And if I didn't have CYGWIN set, I wouldn't get colors either.

Yes, I think you're only getting color because you have `CYGWIN` set.

I don't think the `cmd.exe` shell is involved, though.  In the case where you're using ConEmu, the Cygwin/MSYS2 connector, and Packer, I think the the connector is creating a Cygwin/MSYS2 pty to capture Packer's output, and then the connector copies data from the pty directly to the hook DLL, which I think inteprets the escapes.  ([This line](https://github.com/Maximus5/cygwin-connector/blob/7a1404cc40a43707d9ac9266fbe48e15ce2d827a/ConEmuT.cpp#L363) is probably important.)

Yeah, if Terraform works, then reusing its logic may be a good idea.
